### PR TITLE
feat(outputs.mqtt): Add option to specify topic layouts

### LINKED ISSUE #10732: Extend MQTT output topic formatting options
## Feature Request

The current output options of the MQTT output plugin are limited and have some almost hardcoded assumptions (hostname). I would like to propose an extension suiting the nature of MQTT. 

This proposal is  almost 'the inverse' of #10712. 
### Proposal:

Support something like:

Input:
```
'''
sensors,site=CLE,version=v1,device_name=device5,field=temp1 value=21.0 (float)
sensors,site=CLE,version=v1,device_name=device5,field=spin2 value=100i (int)
sensors,site=CLE,version=v1,device_name=device5,field=status value="OK" (string)
...
```

New proposed type of config:
```
[[output.mqtt_producer]]
    ....
    [[output.mqtt_producer.topic_defintion]]
    topic = "/sensors/_/_/_/_"
    measurement = "/measurement/_/_/_/_"
    tags = "_/site/version/device_name/_"
    fields = "_/_/_/field"
```
Where `topic` specifies the fixed topic parts and defines the variable part the other options have to fill (once and only once).
`measurement` and `tags` map the respective tags and measurement values in the parts of the topic. `field` is replaced with the field name.
 
Will result in MQTT messages output:
```
 Topic                       Message
 /sensors/CLE/v1/device5/temp1 21.0
 /sensors/CLE/v1/device5/temp2 21.0
 /sensors/CLE/v1/device5/rpm     100
 /sensors/CLE/v1/device5/ph       7.0
 /sensors/CLE/v1/device5/spin1   1
 /sensors/CLE/v1/device5/spin2  13
 /sensors/CLE/v1/device5/status OK
 /sensors/CLE/v1/device5/log      INFO: happy
```
(Note: this example is the exact inverse of the example in #10712) A unit test would be to "output what you can consume as input".

### Current behavior:

Currently it would require a lot of seperate MQTT producers to achieve this example.

### Desired behavior:

See example.

### Use case: 

In local IOT systems based on MQTT backbone  InfluxDB is a very powerful producer of aggregated data (e.g. daily, weekly stats) you want to push this data out to MQTT as well. JSON/influxdb parsing is not always a strong point or possible on IOT devices OR you have to adjust to existing patterns on the existing MQTT installation.

### LINKED ISSUE #12589: MQTT.output format optimization
### Use Case

An increasing amount of IoT devices support MQTT these days, however some are not (due to age or design or other reasons). Other popular alternatives to MQTT are JSON http or Modbus TCP for example. 

In a typical smart home or any business environment with many different devices there is a good use case for Telegraf to act as a centralized communications bus (a bit like Kafka), especially since there is already a vast amount of input and output plugins available. This consolidates many different devices like power inverters, thermostats, heating pumps, etc. into a single tool which then can both write the timeseries data to influxdb but also reformat the various inputs to a consistent outplut like for example mqtt which then is easy to consume by other tools, like automation (iobroker, Homeassistant, openHAB, etc..)

My proposal is to optimize the mqtt.output feature such that the resulting mqtt messages are more aligned with common practices for mqtt where you often see a value/key pair having forming it's own (sub-)topic. Currently you would most likely see MQTT messages in a format where you have a topic like 

```
/source-device/type-of-data/key/value
/thermostat_A/room/temperature/18
/thermostat_A/room/humidity/45
```

This would encompassing to take in different key:value data pairs from the input (e.g. modbus) and transfer them into some logic and flexible mqtt topic structure like explained above. 

### Expected behavior

Take input of key:value data pairs and optionally any descriptive tags from incoming source data such as for example modbus or http json and construct an mqtt.output with a format that is easy to understand and parse for any mqtt consumer.

The expected format would be like illustrated above for thermostat that provides for a room temperature and humidity for example:
The topic would for example come from:
device-name (probably form the input plugin device name tag)
room either a tag for this value or optionally inserted by the user in the config file
the last two values would be key and value from the measurement.

```
/thermostat_A/room/temperature/18
/thermostat_A/room/humidity/bedroom/45
```

### Actual behavior

Currently for example in my case the modbus plugin creates a contiguous stream of data (tag key value, ...) which are all posted into a single mqtt topic. for example:

/telegraf/modbus/thermostat/temperature 18, humidity 45

### Additional info

Let me know if you want me to provide an actual example of modbus data and how I'd like it structured.

### PR DESCRIPTION
**Pull request by srebhan**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #12589
resolves #10732 

This PR adds a new `layout` option to specify how topics are generated from metrics. While preserving the original _batch_ and _non-batch_ options (rendering `batch` deprecated), the PR adds two new options, `field` and `homie-v4`. The `field` layout will generate one topic per metric field only containing the field value in string format (solving #10732). The `homie-v4` layout allows to output metrics in topics adhering to the [Homie v4.0 specification](https://homieiot.github.io/specification/spec-core-v4_0_0), a standard used for M2M and home-automation purposes.

### CODE REVIEW

**Comment by srebhan** [comment]
Based on #12683 and thus marked as draft for now.

**Comment by fietser28** [comment]
@srebhan I probably won't have time to setup a test for this the coming week. From the documentation it looks like this will solve my problem in my specific case (single task creating multiple fields that can be translated to separate mqtt topics).

**Comment by srebhan** [comment]
Ok. Let me know if you were able to really test this.

File: plugins/outputs/mqtt/README.md [hunk 1]
```diff
@@ -121,3 +140,194 @@ to use them.
...
+telegraf/modbus/device-2/account-no/$datatype     string
+telegraf/modbus/device-2/supplied                 true
+telegraf/modbus/device-2/supplied/$name           supplied
+telegraf/modbus/device-2/supplied/$datatype       boolean
+```
+
+#### Important notes and limitations
+
+It is important to notice that the __"devices" and "nodes" are dynamically
+changing_ in Telegraf as the metrics and their structure is not known a-priori.
```
Comment (powersj) [inline plugins/outputs/mqtt/README.md:307]:
```suggestion
It is important to notice that the __"devices" and "nodes" are dynamically
changing__ in Telegraf as the metrics and their structure is not known a-priori.
```

File: plugins/outputs/mqtt/homie.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	if _, found := m.homieSeen[topic]; !found {
+		deviceName, err := m.homieDeviceNameGenerator.Generate(metric)
+		if err != nil {
+			return nil, "", fmt.Errorf("generating device name failed: %w", err)
+		}
+		messages = append(messages, message{topic + "/$homie", []byte("4.0")})
+		messages = append(messages, message{topic + "/$name", []byte(deviceName)})
+		messages = append(messages, message{topic + "/$state", []byte("ready")})
+		m.homieSeen[topic] = make(map[string]bool)
+		// TODO: Declare last will
```
Comment (powersj) [inline plugins/outputs/mqtt/homie.go:30]: Is there a follow-up PR needed here? Or something we can do to document this TODO?

**Review by powersj — COMMENTED** [review]
couple inline questions, otherwise this looks good to go! Thanks!

**Review by powersj — APPROVED** [review]

**Pull request merged**